Invitation: Display responses and allow the local calendar to be updated

NeilRashbrook opened the pull request:
Still to do:
- [ ] Send invitations
  - [ ] IMAP
  - [ ] ActiveSync
  - [X] EWS
  - [X] OWA
- [X] Recognise invitation email
- [X] Display invitation data
- [X] Accept or decline invitation
- [X] Recognise invitation event
- [X] Change invitation acceptance
- [X] Recognise acceptance email
- [X] Update meeting participation

benbucksch commented on app/frontend/Calendar/DisplayEvent/Invitation.svelte:
```diff
@@ -14,88 +14,67 @@
-          icon={MaybeIcon}
-          classes="maybe" />
-      </ButtonMenu>
-    </hbox>
```
Please keep the UI as-is. We had intentionally put this into a menu.

NeilRashbrook replied:
This was the result of a merge conflict. At least I got the right label.

benbucksch replied:
So, could you please restore it?

benbucksch replied:
Please also check whether there were other "merge errors".

benbucksch commented on app/frontend/Calendar/DisplayEvent/Invitation.svelte:
```diff
@@ -14,88 +14,67 @@
-          classes="maybe" />
-      </ButtonMenu>
-    </hbox>
-  {/if}
```
We want to show how our user answered to a given invitation email. Please keep the UI.

NeilRashbrook replied:
Sadly it doesn't actually work, at least, not in the state it's in right now. Leaving it here would also break the fix I'm trying to make. And fixing it is beyond the scope of this PR.

benbucksch replied:
> Leaving it here would also break the fix I'm trying to make.

How it would break with this PR, and why?

We've worked on this UI quite a lot. In fact, 3 people in the team worked together on this. The translated labels were pain to get right, and needed several rounds to be correctly translated in various languages. I do no want that work to get lost. Please consulting with me before making such drastic changes to the UI.

At the very least, please wrap them in an `{#if false}`, and then follow up with another PR to fix it and make them functional.

benbucksch replied:
FWIW, the UI specs are here: https://parula.cloud.xwiki.com/xwiki/bin/view/Main/Calendar/Email%20invitation/

NeilRashbrook replied:
Because Scheduling.Accepted and Scheduling.Declined have always referred to incoming messages from attendees that you invited and have nothing whatsoever to do with how you responded to an invitation.

benbucksch replied:
Ah, OK, then I misunderstood the API. We need to straighten that out, to make it clearer.
`{#if false}` is fail, then. Thanks that you did that and restored the buttons.

benbucksch commented on app/frontend/Calendar/DisplayEvent/InvitationDisplay.svelte:
```diff
@@ -6,7 +6,6 @@
       {/if}
     </hbox>
     <hbox flex />
-    <AccountDropDown selectedAccount={calendar} accounts={appGlobal.calendars} withIcon={true} withLabel={false} />
```
The user can select which calendar to put the event in. That is particularly needed for IMAP account in combination with multiple local or CalDAV calendars.

NeilRashbrook replied:
It's not gone, it just got moved somewhere where it was easier for me to populate it correctly.

benbucksch reviewed (changes requested):
This PR is removing significant parts of the UI had we need.

Also the APIs are not right yet. As mentioned, I would like the Invitation handling functions on a separate OutgoingInvitation (and IncomingInvitation) class.

NeilRashbrook reviewed (commented):
> Also the APIs are not right yet. As mentioned, I would like the Invitation handling functions on a separate OutgoingInvitation (and IncomingInvitation) class.

I thought about that, but I was too tired at the time to refactor it.

NeilRashbrook commented:
> I would like the Invitation handling functions on a separate class.

I've refactored them again. And while I was there, I mostly fixed the invitation UI, so that for IMAP accounts you can see how you responded to the invitation (if it's still in your calendar), even after the fact.

benbucksch commented:
Thank you!

benbucksch commented on app/logic/Calendar/IncomingActions.ts:
```diff
@@ -0,0 +1,64 @@
+import type { EMail } from "../Mail/EMail";
+import { assert } from "../util/util";
+
+export class IncomingActions {
```
Please rename to `IncomingInvitation` and move to app/logic/Calendar/Invitation/IncomingInvitation.ts . Ditto for `ActiveSyncIncomingInvitation` etc.

benbucksch commented on app/logic/Calendar/IncomingActions.ts:
```diff
@@ -0,0 +1,64 @@
+    /* else add participant? */
+  }
+
+  async updateFromResponse() {
```
The function name is not clear to me. Are these responses from other participants (or the organizer) about whether the other people are going to attend? If so, can you rename it to `updateFromOtherParticipant()` ?

NeilRashbrook replied:
It actually handles two cases:
- Cancellations from meeting organisers from meetings they invited you to
- Replies from participants to meetings you organised

But then again you knew I would name it badly...

NeilRashbrook commented:
I've rebased everything onto master again, as there were a boatload of merge conflicts, and while I was there I also addressed your review comments.

benbucksch commented on app/frontend/Calendar/DisplayEvent/Invitation.svelte:
```diff
@@ -14,88 +14,119 @@
+        </ButtonMenu>
+      {/if}
+    {:else if message.invitationMessage}
+      <Button label={$t`Update calendar`} disabled={updateDisabled} onClick={onUpdate} classes="font-normal" />
```
I don't understand what this button does. (And if I don't understand, how will a user understand?)

benbucksch replied:
Can you please do that automatically and remove the button?

benbucksch replied:
Good. Nice. Do that.

benbucksch replied:
Please do that in a new PR, so that this here can land.

benbucksch replied:
Filed #619

benbucksch commented on app/frontend/Calendar/DisplayEvent/Invitation.svelte:
```diff
@@ -14,88 +14,119 @@
+  }
+  function selectCalendar() {
+    incomingInvitation = selectedCalendar.getIncomingInvitationFor(message);
+    myResponse = incomingInvitation.myResponse;
```
This is strange. 

The account selector is there to allow the user to select which calendar the meeting will be placed in, at the moment when he clicks Accept. Once the user selected a response, we show that state when the user clicks on the message again (quite possibily on another day).

Am I reading that right that it will clear the response display once the user changed the calendar in the calendar selector? That's confusing, because the user might think that it cleared his response at the organizer, but it did not.

benbucksch replied:
Can we move the event to the other calendar, using `event.moveToCalendar()`?
If that is much more difficult, then disable the calendar selector after response, yes.

NeilRashbrook replied:
I can look into that.

benbucksch commented on app/logic/Mail/EMail.ts:
```diff
@@ -201,36 +202,11 @@ export class EMail extends Message {
-    await this.folder.account.sendInvitationResponse(this.event, response);
+  getUpdateCalendars(): Collection<Calendar> {
+    assert(this.invitationMessage && this.event, "Must have event to find calendar");
+    let validCalendars = appGlobal.calendars.filter(calendar => !calendar.mainAccount);
```
Why this condition that the calendar cannot be related to an email account?

Note that we will have CalDAV accounts that will have an IMAP account as mainAccount.

benbucksch commented on app/logic/Mail/EMail.ts:
```diff
@@ -201,36 +202,11 @@ export class EMail extends Message {
+  getUpdateCalendars(): Collection<Calendar> {
+    assert(this.invitationMessage && this.event, "Must have event to find calendar");
+    let validCalendars = appGlobal.calendars.filter(calendar => !calendar.mainAccount);
+    let foundCalendars = validCalendars.filter(calendar => calendar.events.some(event => event.calUID == this.event.calUID));
```
It seems that you're primarily looking for the calendar that has this event. If so, why not return the Event in the Calendar directly?

(validCalendars looks bogey to me anyways.)

If we can move this function somewhere into Calendar/, that would be better.

benbucksch commented on app/logic/Mail/EWS/EWSAccount.ts:
```diff
@@ -31,6 +31,7 @@ export class EWSAccount extends MailAccount {
   readonly tls = TLSSocketType.TLS;
   readonly canSendInvitations: boolean = false;
   readonly folderMap = new Map<string, EWSFolder>;
+  calendar: EWSCalendar | void;
```
Can you please remove this?

I'm trying to keep the zig-zack dependencies at bay. Email should not depend on Calendar.

(The code below that creates the calendar should also eventually move out of here. E.g. into setup. One hack doesn't justify another.)

benbucksch replied:
Yes, please. Shared calendars are coming right after.

benbucksch reviewed (changes requested):
This is going in the right direction, but not there yet.

benbucksch commented:
Please continue work on CardDAV for now.

benbucksch commented on app/frontend/Shared/Button.svelte:
```diff
@@ -89,7 +89,9 @@
       loading = false;
     }
-    disabled = previousDisabled;
+    if (disabled === true) {
```
`previousDisabled` was supposed to restore it to the previous state before onClick ran, whether it's disabled or enabled.

The only condition that I see where your change make a difference is if the `disabled` Svelte property changed while the onClick command was running.

NeilRashbrook replied:
Well, yes, I wanted the callback to be able to keep the button disabled.

benbucksch replied:
So, the `onClick` function changes it to disabled. OK, makes sense.

benbucksch commented on app/frontend/Calendar/EditEvent/InvitationResponseButtons.svelte:
```diff
@@ -1,5 +1,5 @@
 <hbox class="buttons">
-  {#if event.response == InvitationResponse.Unknown || event.response == InvitationResponse.Organizer}
+  {#if event.response > InvitationResponse.Organizer && event.participants.some(participant => participant.response == InvitationResponse.Organizer)}
```
`>`: Please avoid such order comparisons for enums. See #534

benbucksch commented on app/logic/Mail/OWA/OWAAccount.ts:
```diff
@@ -34,6 +34,7 @@ export class OWAAccount extends MailAccount {
   readonly tls = TLSSocketType.TLS;
   readonly canSendInvitations: boolean = false;
   readonly folderMap = new Map<string, OWAFolder>;
+  calendar: OWACalendar | void;
```
ditto

NeilRashbrook commented:
(Converted to draft as this temporarily depends on PR #604)

benbucksch commented:
Hey Neil, could you please now come back to this and finish this, addressing the review comments? Also, adapt to the changes that happened on master, including the renames etc. I would like to land this tomorrow (Tuesday), if possible.

benbucksch commented on app/frontend/Calendar/DisplayEvent/InvitationDisplay.svelte:
```diff
@@ -6,22 +6,26 @@
     <hbox flex />
-    <AccountDropDown selectedAccount={calendar} accounts={appGlobal.calendars} withIcon={true} withLabel={false} />
+    {#if selectedCalendar}
+      <AccountDropDown selectedAccount={selectedCalendar} accounts={calendars} withIcon={true} withLabel={true} disabled={calendars.length < 2} on:Select={selectCalendar} />
```
Simply don't show the dropdown, if there's only 1 calendar.

benbucksch commented on app/frontend/Shared/AccountDropDown.svelte:
```diff
@@ -24,6 +24,7 @@
   export let accounts: Collection<Account>;
   export let withIcon: boolean = false;
   export let withLabel: boolean = true;
+  export let disabled: boolean = false;
```
Not needed, simply don't show, see above.

benbucksch commented on app/logic/Calendar/Calendar.ts:
```diff
@@ -1,20 +1,27 @@
 export class Calendar extends Account {
   readonly protocol: string = "calendar-local";
   readonly events = new ArrayColl<Event>();
+  readonly canAcceptAnyInvitation: boolean = true;
```
Please document what you mean with "any".

benbucksch commented on app/logic/Calendar/Invitation/IncomingInvitation.ts:
```diff
@@ -0,0 +1,65 @@
+    /* else add participant? */
+  }
+
+  /** Handle the other two InvitationMessage values, as
+   *  respondToInvitation only handles InvitationMessage.Invitation */
```
```suggestion
  /** It handles two cases:
   * - Cancellations from meeting organisers for `IncomingInvitation`
   * - Replies from participants in response to `OutgoingInvitation`
   *   TODO Move second case to `OutgoingInvitation`
   */
```

NeilRashbrook commented:
I've fixed two bugs in the changes that I'd made (it's always possible there are more but ActiveSync keeps timing out on me and waiting 20 minutes would not be good use of my time) and I've also cherry-picked your commit a6c2cb6 and then I've force-pushed my branch (because I've already rebased it).

benbucksch commented on app/logic/Calendar/Calendar.ts:
```diff
@@ -1,20 +1,28 @@
 export class Calendar extends Account {
   readonly protocol: string = "calendar-local";
   readonly events = new ArrayColl<Event>();
+  /** Can this calendar accept incoming invitations from any source? */
```
```suggestion
  /** Can this calendar accept incoming invitations from inboxes in other accounts? */
```

benbucksch commented on app/logic/Calendar/Invitation/IncomingInvitation.ts:
```diff
@@ -0,0 +1,63 @@
+        await event.save();
+      }
+    } else {
+      let invitee = this.event.participants.find(participant => participant.response != InvitationResponse.Organizer);
```
You're assuming that there will be only 2 participants in the response: The organizer and the sender. Is that mandated by the spec? If not, it's conceivable that there is some software that sends *all* participants. (Even if you have not actually encountered such a message. We have not seen all software that sends iCal.)

I think there should be an extra check that either a) there are <= 2 participants in the list or b) the participant email address matches the From email address of the email.

NeilRashbrook replied:
in theory there is apparently something called delegation - you receive an invitation, but you want someone else to take your place. In this case you're supposed to send the organiser a message that you want to delegate the invitation and the other person sends to both people whether they can attend, and the resulting message has three participants.

Unfortunately whenever I try to search for support for this feature I only get the completely different feature of Office 365/Exchange whereby you can look at another user's calendar, inbox etc.

benbucksch replied:
Yes, I don't think Exchange does it this way.

benbucksch replied:
I was thinking about the case where some calendar client on the other end simply copies *all* participants from the original invitation that they received, copied all their status, and changed only their own status. I am not sure such a software exists, but it's conceivable, and we should not break in this case.

benbucksch replied:
Also, this needs a null check. If there is no other participant is found, `invitee` will be null, and `invitee.emailAddress` and `invitee.response` below will throw.

NeilRashbrook replied:
Given that, how should we notify the user that the reply is malformed?

benbucksch replied:
I guess the best would be an inline error message inside the invitation display.
